beacon_node/network/src/sync/block_lookups/mod.rs

```diff
@@ -1,101 +1,272 @@
+        );
+    }
+
+    pub fn search_child_block(
+        &mut self,
+        block_root: Hash256,
+        block: Option<Arc<SignedBeaconBlock<T::EthSpec>>>,
+        blobs: Option<FixedBlobSidecarList<T::EthSpec>>,
+        peer_source: &[PeerShouldHave],
+        cx: &mut SyncNetworkContext<T>,
+    ) {
+        self.search_block_with(
+            block_root,
+            block,
+            blobs,
+            peer_source,
+            LookupSource::UnknownParent,
+        );
+        self.trigger_single_lookup(block_root, cx)
+    }
+
+    pub fn search_child_delayed(
+        &mut self,
+        block_root: Hash256,
+        block: Option<Arc<SignedBeaconBlock<T::EthSpec>>>,
+        blobs: Option<FixedBlobSidecarList<T::EthSpec>>,
+        peer_source: &[PeerShouldHave],
+    ) {
+        self.search_block_with(
+            block_root,
+            block,
+            blobs,
+            peer_source,
+            LookupSource::UnknownParent,
+        );
+    }
+
+    /// Attempts to trigger the request matching the given `block_root`. If the requests for
+    /// blocks and blobs could not have been made or are no longer required, this also removes the lookup.
+    pub fn trigger_single_lookup(&mut self, block_root: Hash256, cx: &mut SyncNetworkContext<T>) {
+        let mut drop_lookup = None;
+        for (index, single_block_request) in self.single_block_lookups.iter_mut().enumerate() {
+            if single_block_request
+                .block_request_state
+                .requested_block_root
+                == block_root
+                && !single_block_request.triggered
+            {
+                if single_block_request.request_block_and_blobs(cx).is_err() {
+                    drop_lookup = Some(index);
+                }
+                single_block_request.triggered = true;
+            }
+        }
+        if let Some(index) = drop_lookup {
+            self.single_block_lookups.swap_remove(index);
+        }
+    }
+
     /// Searches for a single block hash. If the blocks parent is unknown, a chain of blocks is
     /// constructed.
-    pub fn search_block(&mut self, hash: Hash256, peer_id: PeerId, cx: &mut SyncNetworkContext<T>) {
+    pub fn search_block_with(
+        &mut self,
+        block_root: Hash256,
+        block: Option<Arc<SignedBeaconBlock<T::EthSpec>>>,
+        blobs: Option<FixedBlobSidecarList<T::EthSpec>>,
+        peers: &[PeerShouldHave],
+        lookup_source: LookupSource,
+    ) {
         // Do not re-request a block that is already being requested
-        if self
+        if let Some(lookup) = self
             .single_block_lookups
-            .values_mut()
-            .any(|single_block_request| single_block_request.add_peer(&hash, &peer_id))
+            .iter_mut()
+            .find(|lookup| lookup.is_for_block(block_root))
         {
+            lookup.add_peers(peers);
+            if let Some(block) = block {
+                lookup.add_unknown_parent_block(block)
```

Comment:
Wouldn't doing this add an `UnknownParentComponent` to a SingleBlockLookup that has `LookupSource::MissingComponents`

Reply:
Ohh this can't happen because `block` and `blobs` both are `None`.

Reply:
yea i think this could be a lot more clear by just passing in `Option<UnknownParentComponents>` as an arg to these methods

Reply:
This actually made me realize `LookupSource` ends up being redundant because the presence or absence of `UnknownParentComponents` ends up giving us the same information

Reply:
Yep, loving the changes 🔥